### PR TITLE
ENH: compressed layout

### PR DESCRIPTION
**Pull request by jklymak**
## PR Summary

Replaces #20016

We _often_ want a grid of fixed-aspect ratio axes, but constrained_layout and tigt_layout leave large blank spaces between the axes because they only work with the original axes positions, not the aspect-modified versions.  

Here I proposed an extension to constrianed_layout that works for simple grids of axes...

### Old:
Exampleconstrained

### New:
Examplecompressed

### Code

```python
import matplotlib.pyplot as plt
import numpy as np

nrows = 2
ncols = 3
for compress in ['constrained', 'compressed']:
    print('compress?', compress)
    fig, axs = plt.subplots( nrows, ncols, figsize=(4,4),
                            layout=compress,
                            sharex=True, sharey=True)
    fig.patch.set_facecolor("0.9")
    #fig.set_constrained_layout_pads(compress=True)

    for i in range(nrows):
        for j in range(ncols):
            ax = axs[i, j]
            ax.set_aspect(1)
            pc = ax.pcolormesh(np.random.randn(30, 30))
    axs[0, 1].set_xlabel('Boooo')
    fig.colorbar(pc, ax=axs)
```

### Failures...

This isn't perfect, and that is why it is not in "constrained_layout"   Folks should be able to turn this off...  This example isn't terrible, but its not very good either, though

```python
fig, axs = plt.subplot_mosaic([['A', 'B', 'B'], ['C', 'D', 'E']], layout='compressed',
                            sharex=True, sharey=True)
for k in axs:
    axs[k].imshow(X)
```

ExampleKindaBad

Or:

```python
fig, axs = plt.subplot_mosaic([['A', 'B', 'B'], ['C', 'B', 'B']], layout='compressed',
                            sharex=True, sharey=True, figsize=(3, 7))

```
Better

## PR Checklist

**Tests and Styling**
- [ ] Has pytest style unit tests (and `pytest` passes).
- [ ] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [ ] New features are documented, with examples if plot related.
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

### CODE REVIEW

File: lib/matplotlib/layout_engine.py
Comment (tacaswell) [inline lib/matplotlib/layout_engine.py:208]: This comment seems wrong now?

Comment (jklymak) [reply lib/matplotlib/layout_engine.py]: Ooops, yes, I decided against needing a subclass here.

File: lib/matplotlib/_constrained_layout.py
Comment (QuLogic) [inline lib/matplotlib/_constrained_layout.py:274]: I think you can do this with NumPy?
```suggestion
                extraw[gs][sub.colspan] = np.maximum(extraw[gs][sub.colspan], dw)
```

File: lib/matplotlib/_constrained_layout.py
Comment (QuLogic) [inline lib/matplotlib/_constrained_layout.py:284]: Remove extra sums
```suggestion
    w = np.sum(extraw[gs]) / 2
    layoutgrids[fig].edit_margin_min('left', w)
    layoutgrids[fig].edit_margin_min('right', w)

    h = np.sum(extrah[gs]) / 2
    layoutgrids[fig].edit_margin_min('top', h)
    layoutgrids[fig].edit_margin_min('bottom', h)
```

File: tutorials/intermediate/constrainedlayout_guide.py
Comment (jklymak) [reply tutorials/intermediate/constrainedlayout_guide.py:460]: Changed to the "original position" which is the terminology we use in `ax.get_position`:

```
# By default, ``constrained_layout`` acts on the original position of an axes.
# However, when Axes have fixed aspect ratios, one side is made shorter.  For
# a grid of such Axes, this can leave a large gap in the shortened direction.
# In the following the Axes are sqaure, but the figure quite wide so there is
# a horizontal gap:
```

File: lib/matplotlib/_constrained_layout.py
Comment (jklymak) [reply lib/matplotlib/_constrained_layout.py:274]: That seems to work.  OTOH, I'm not sure its that readable?

File: lib/matplotlib/_constrained_layout.py
Comment (QuLogic) [inline lib/matplotlib/_constrained_layout.py:280]: It appears that only the last `gs` is used here; is it assumed to be the same `gs` for every `ax`? If that's the case, then why do `extraw`/`extrah` need to be dictionaries?

Comment (jklymak) [reply lib/matplotlib/_constrained_layout.py:280]: Good catch.  I probably had dreams of something fancier here.  I guess th right thing is to error out if gs changes...

File: tutorials/intermediate/constrainedlayout_guide.py
Comment (timhoffm) [inline tutorials/intermediate/constrainedlayout_guide.py:464]: Is "original position of an axes" a concept we have? Maybe

 ```suggestion
# By default, ``constrained_layout`` acts on the grid spaces reserved for the
# Axes. However, when Axes have fixed aspect ratios, one of their sides is made
# shorter. so that it fits the reserved space while ensuring the fixed aspect
# ratio. A grid of such Axes, may thus have large gaps between the axes:
```

File: tutorials/intermediate/constrainedlayout_guide.py
Comment (timhoffm) [inline tutorials/intermediate/constrainedlayout_guide.py:470]: Using the keyword notation may make the connection the code even stronger:
```suggestion
fig.suptitle("fixed-aspect plots, layout='constrained'")
```

File: tutorials/intermediate/constrainedlayout_guide.py
Comment (timhoffm) [inline tutorials/intermediate/constrainedlayout_guide.py:480]:
```suggestion
fig.suptitle("fixed-aspect plots, layout='compressed'")
```

File: tutorials/intermediate/constrainedlayout_guide.py
Comment (jklymak) [reply tutorials/intermediate/constrainedlayout_guide.py:202]: If you run this standalone, this directory will likely not exist...   We need to show the figure with bbox_inches='tight' which I don't think we can do with sphinx gallery?

Comment (timhoffm) [reply tutorials/intermediate/constrainedlayout_guide.py:202]: sphinx-gallery: You've already discussed that and it's not supported: https://github.com/sphinx-gallery/sphinx-gallery/issues/726 :smile: 

I'm inclined to let the user stumble over this and fix the path if they really run this part. Readability and simplicity > automatic error correction.

If you really would want to keep this, I think we'd need to comment on the error. As is this is quite confusing. OTOH commenting would make this even more verbose and distract from the actual example.

File: lib/matplotlib/_constrained_layout.py
Comment (QuLogic) [inline lib/matplotlib/_constrained_layout.py:268]: This seems unused before getting re-assigned at line 280.

File: lib/matplotlib/_constrained_layout.py
Comment (QuLogic) [inline lib/matplotlib/_constrained_layout.py:299]: Would flip this to `if gs is None: raise` and dedent the rest.

File: tutorials/intermediate/constrainedlayout_guide.py
Comment (QuLogic) [inline tutorials/intermediate/constrainedlayout_guide.py:472]: Needs the `####...` header if you want this to be text instead of a comment.

File: tutorials/intermediate/constrainedlayout_guide.py
Comment (QuLogic) [inline tutorials/intermediate/constrainedlayout_guide.py:478]: This seems to look the same as the constrained version?

File: tutorials/intermediate/constrainedlayout_guide.py
Comment (jklymak) [reply tutorials/intermediate/constrainedlayout_guide.py]: That's really weird.  If I do this manually it works fine.  if I save it, it works fine.  However, I agree that the sphinx version does not work fine?

File: tutorials/intermediate/constrainedlayout_guide.py
Comment (jklymak) [reply tutorials/intermediate/constrainedlayout_guide.py:479]: Agreed, https://output.circle-artifacts.com/output/job/4eee1b1a-1268-4881-8d21-61000c011c8e/artifacts/0/doc/build/html/tutorials/intermediate/constrainedlayout_guide.html#grids-of-fixed-aspect-ratio-axes-compressed-layout

but I have no idea what is happening here.  This example works fine manually, just using Agg, etc.  I'm flummoxed why the sphinx version doesn't work...

Comment (jklymak) [reply tutorials/intermediate/constrainedlayout_guide.py:480]: Fixed in newest version - there was some weird interactions between the rcParam and subsequent calls.  maybe worth chasing down, but I could not reproduce without sphinx-gallery, so I guess sphinx-galery is using some weird rcParam contexts?

Comment (jklymak) [reply tutorials/intermediate/constrainedlayout_guide.py:480]: See https://output.circle-artifacts.com/output/job/a187bd70-81b0-4ede-ae1c-ab250a32d50c/artifacts/0/doc/build/html/tutorials/intermediate/constrainedlayout_guide.html#grids-of-fixed-aspect-ratio-axes-compressed-layout

File: tutorials/intermediate/constrainedlayout_guide.py
Comment (QuLogic) [inline tutorials/intermediate/constrainedlayout_guide.py:472]: Extra space
```suggestion
# One obvious way of fixing this is to make the figure size more square,
```

File: tutorials/intermediate/constrainedlayout_guide.py
Comment (QuLogic) [reply tutorials/intermediate/constrainedlayout_guide.py:480]: I think the context remains the same through a single file.

File: lib/matplotlib/_constrained_layout.py
Comment (QuLogic) [inline lib/matplotlib/_constrained_layout.py:267]: I suppose this could also flip to `if not hasattr(ax, 'get_subplotspec'): continue` as well.

**Review by QuLogic — APPROVED** [review]

File: lib/matplotlib/_constrained_layout.py
Comment (oscargus) [inline lib/matplotlib/_constrained_layout.py:87]: For consistency with above, this empty line should be here (and one added after the `rect` description) or all above also removed.

File: tutorials/intermediate/constrainedlayout_guide.py
Comment (jklymak) [reply tutorials/intermediate/constrainedlayout_guide.py:480]: Yes, but even if I change the rcParam in a normal script, and then specify the kwarg, `compress` works fine.  But in sphinx- gallery it has this weird problem.

File: tutorials/intermediate/constrainedlayout_guide.py
Comment (jklymak) [reply tutorials/intermediate/constrainedlayout_guide.py:202]: Sorry to drop this conversation - _I_ tripped across this trying to debug the tutorial.  I will add a comment...

File: lib/matplotlib/_constrained_layout.py
Comment (jklymak) [reply lib/matplotlib/_constrained_layout.py:267]: Done.  At some point we were taught that "break" and "continue" are not desirable.

File: lib/matplotlib/_constrained_layout.py
Comment (jklymak) [reply lib/matplotlib/_constrained_layout.py:87]: Sure, though this is private, so never rendered in the docs ;-)

**Comment by jklymak** [comment]
I've marked merge with single review.  This is pretty stand alone, so it shouldn't break anyone.  However, happy to get a second review.  I'll merge by next Friday if it doesn't attract one.

**Review by timhoffm — APPROVED** [review]

**Comment by jklymak** [comment]
Thanks everyone for your work on this!

**Comment by timhoffm** [comment]
> Thanks everyone for your work on this!

Thanks for creating this! :smile:

**Pull request merged**
